privacyidea/lib/utils/__init__.py

```diff
@@ -1480,3 +1481,41 @@ def parse_string_to_dict(s, split_char=":"):
     values = [[x for x in y.split()] for y in packed_list[1::2]]
     d = {a: b for a, b in zip(keys, values)}
     return d
+
+
+def replace_function_event_handler(text, token_serial=None, tokenowner=None, logged_in_user=None):
+    if logged_in_user is not None:
+        login = logged_in_user.login
+        realm = logged_in_user.realm
+    else:
+        login = ""
+        realm = ""
+
+    if tokenowner is not None:
+        surname = tokenowner.info.get("surname")
+        givenname = tokenowner.info.get("givenname")
+        userrealm = tokenowner.realm
+    else:
+        surname = ""
+        givenname = ""
+        userrealm = ""
+
+    if token_serial is not None:
+        token_serial = token_serial
+    else:
+        token_serial = ""
+
+    try:
+        attributes = {
+            "logged_in_user": login,
+            "realm": realm,
+            "surname": surname,
+            "token_owner": givenname,
+            "user_realm": userrealm,
+            "token_serial": token_serial
+        }
+        new_text = text.format(**attributes)
+        return new_text
+    except(ValueError, KeyError) as err:
+        log.warning('privacyIDEA can`t replace your placeholder. Please check the text for mistakes')
```

Comment:
You can add the error message to the warning, makes it easier to figure out, which placeholder was wrong.

Reply:
maybe also add a 

    log.debug(traceback...))

Reply:
09f984a